artiq/firmware/runtime/sched.rs

```diff
@@ -250,14 +250,26 @@ impl<'a> Io<'a> {
         })
     }
 
-    pub fn until<F: FnMut() -> bool>(&self, mut f: F) -> Result<(), Error> {
+    pub fn inner_until<F: FnMut() -> bool>(
```

Comment:
Wouldn't it be simpler to flush and close the socket in the core analyzer server? This way the ``Io`` abstraction layer would not deviate from the smoltcp API.

Reply:
That would work, I didn't have confidence that that would be the only place where this was needed, or that it wouldn't be a paper cut for others adding more features in future. I was preferring simplicity of use over simplicity of implementation. 

Whilst we're discussing this I think there's a third possible design here: we could have a task that owns closing sockets when a socket is dropped if it hasn't already been closed it could be added to a list of sockets so that that task can flush it and close it down. That'll make the Drop implementation simpler and mean that Drop won't block. But might be considered more complex over all.

Reply:
I tend to prefer explicit flush in the application - this is also how it is with a number of existing APIs such as Python asyncio.

Reply:
Sure, I'll make that change then